swarm/storage/mru/lookup.go

```diff
@@ -17,96 +17,94 @@
 package mru
 
 import (
-	"encoding/binary"
 	"fmt"
 	"hash"
-	"net/url"
 	"strconv"
 
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/swarm/storage/mru/lookup"
+
 	"github.com/ethereum/go-ethereum/swarm/storage"
 )
 
 // LookupParams is used to specify constraints when performing an update lookup
 // Limit defines whether or not the lookup should be limited
 // If Limit is set to true then Max defines the amount of hops that can be performed
 type LookupParams struct {
-	UpdateLookup
-	Limit uint32
-}
-
-// Values interface represents a string key-value store
-// useful for building query strings
-type Values interface {
-	Get(key string) string
-	Set(key, value string)
+	View
+	Hint      lookup.Epoch
+	TimeLimit uint64
 }
 
 // FromValues deserializes this instance from a string key-value store
 // useful to parse query strings
-func (lp *LookupParams) FromValues(values Values, parseView bool) error {
-	limit, _ := strconv.ParseUint(values.Get("limit"), 10, 32)
-
-	lp.Limit = uint32(limit)
-	return lp.UpdateLookup.FromValues(values, parseView)
+func (lp *LookupParams) FromValues(values Values) error {
+	time, _ := strconv.ParseUint(values.Get("time"), 10, 64)
+	lp.TimeLimit = uint64(time)
+
+	level, _ := strconv.ParseUint(values.Get("hint.level"), 10, 32)
+	lp.Hint.Level = uint8(level)
+	lp.Hint.Time, _ = strconv.ParseUint(values.Get("hint.time"), 10, 64)
+	if lp.View.User == (common.Address{}) {
+		return lp.View.FromValues(values)
+	}
+	return nil
 }
 
 // ToValues serializes this structure into the provided string key-value store
 // useful to build query strings
-func (lp *LookupParams) ToValues(values url.Values) {
-	if lp.Limit != 0 {
-		values.Set("limit", fmt.Sprintf("%d", lp.Version))
+func (lp *LookupParams) ToValues(values Values) {
+	if lp.TimeLimit != 0 {
+		values.Set("time", fmt.Sprintf("%d", lp.TimeLimit))
 	}
-	lp.UpdateLookup.ToValues(values)
+	if lp.Hint.Level != 0 {
+		values.Set("hint.level", fmt.Sprintf("%d", lp.Hint.Level))
+	}
+	if lp.Hint.Time != 0 {
+		values.Set("hint.time", fmt.Sprintf("%d", lp.Hint.Time))
+	}
+	lp.View.ToValues(values)
 }
 
-// NewLookupParams constructs a LookupParams structure with the provided lookup parameters
-// if period = 0 and version = 0, the last resource version will be looked up
-// if period !=0 and version = 0, the last version in that period will be looked up
-// if both period, version !=0, that specific period, version will be looked up
-func NewLookupParams(view *View, period, version uint32, limit uint32) *LookupParams {
+// LookupBefore constructs an UpdateLookup structure to find updates on or before `time`
+// if time == 0, the latest update will be looked up
+func LookupBefore(view *View, time uint64, hint lookup.Epoch) *LookupParams {
```

Comment:
Just call this `NewHisotryLookupParams`.
Why do we expose this and not just hide it in the actual exported lookup functions?

Reply:
Renamed
There is only one lookup function: `Lookup`, that behaves differently depending on the `LookupParams` you pass.
These functions allow the user to express his intentions clearly.
It also has helped not having to refactor lots of intermediate functions every time we wanted to change implementations.